server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java

```diff
@@ -140,27 +142,50 @@ private MappedFieldType getDynamicField(String field) {
     }
 
     /**
-     * Returns all the mapped field types.
+     * Returns all the mapped field types that match a pattern
+     *
+     * Note that if a field is aliased and both its actual name and its alias
+     * match the pattern, the returned collection will contain the field type
+     * twice.
      */
-    Collection<MappedFieldType> get() {
-        return fullNameToFieldType.values();
+    Collection<MappedFieldType> getMatchingFieldTypes(String pattern) {
+        if ("*".equals(pattern)) {
+            return fullNameToFieldType.values();
+        }
+        if (Regex.isSimpleMatchPattern(pattern) == false) {
+            // no wildcards
+            MappedFieldType ft = get(pattern);
+            return ft == null ? Collections.emptySet() : Collections.singleton(ft);
+        }
+        List<MappedFieldType> matchingFields = new ArrayList<>();
+        for (String field : fullNameToFieldType.keySet()) {
+            if (Regex.simpleMatch(pattern, field)) {
+                matchingFields.add(fullNameToFieldType.get(field));
+            }
+        }
+        return matchingFields;
     }
 
     /**
-     * Returns a list of the full names of a simple match regex like pattern against full name and index name.
+     * Returns a set of field names that match a regex-like pattern
+     *
+     * All field names in the returned set are guaranteed to resolve to a field
      */
-    Set<String> simpleMatchToFullName(String pattern) {
+    Set<String> getMatchingFieldNames(String pattern) {
+        if ("*".equals(pattern)) {
+            return fullNameToFieldType.keySet();
+        }
         if (Regex.isSimpleMatchPattern(pattern) == false) {
             // no wildcards
-            return Collections.singleton(pattern);
+            return get(pattern) == null ? Collections.emptySet() : Collections.singleton(pattern);
```

Comment:
I think this changes the semantics slightly -- if you pass in a flattened subfield `my_flattened.subfield`, we'll that same value, whereas before we didn't. But if you pass in `my_flattened.sub*`, nothing will be returned. I think we should maintain the old behavior, which is more consistent.

Reply:
++ good catch, I'll change it back

Reply:
Turns out we do need this behaviour for flattened queries to work with query_string queries, and the inconsistent behaviour, although a bit odd, is necessary - and I think is what was happening before as well?

Reply:
I managed to confuse myself earlier, I agree the current behavior is correct. The 'inconsistent behavior' I noted is actually just how flattened field resolution works, including `getMatchingFieldTypes`. We support requesting a flattened subfield directly, but do not attempt to resolve wildcard patterns to subfields (this is an explicit choice and noted in docs).